transformers_cfg/parser.py

```diff
@@ -10,22 +13,163 @@
 END_OF_GRAMMAR_MARKER = 0xFFFF
 TO_BE_FILLED_MARKER = 0
 REF_RULE_MARKER = 1
-LITERAL_MARKER = 2
 
 
 ########################
 # EBNF Grammar Parsing #
 ########################
 
 
+class Codable(ABC):
+    def serialize(self) -> List[int]:
+        raise NotImplementedError()
+
+    @classmethod
+    def deserialize(cls, data: List[int]) -> "Codable":
+        raise NotImplementedError()
+
+
+@dataclass
+class GrammarElement(Codable):
+    def is_terminated(self) -> bool:
+        raise NotImplementedError()
+    
+
+@dataclass
+class TerminatedElement(GrammarElement):
+    ranges: List[Tuple[int, int]]
+
+    def is_terminated(self) -> bool:
+        return True
+    
+    def serialize(self) -> List[int]:
+        outbuf = [len(self.ranges) * 2]
+        for range in self.ranges:
+            outbuf.extend(range)
+        return outbuf
+    
+    @classmethod
+    def from_range(cls, start: int, end: int) -> "GrammarElement":
+        return cls([(start, end)])
+
+
+@dataclass
+class ReferenceElement(GrammarElement):
+    reference_id: int
```

Comment:
Maybe the name `referee_id` would be more explictly expressing that this is the symbol id that is referred to